Generated design picker: Sticky thumbnails

arthur791004 opened the pull request:
#### Changes proposed in this Pull Request

* As the WordPress logo is inside `SignupHeader` component and it's not easy to stick at the top of the screen, I make it become an `@automatitc/components` and render it inside the `ActionButtons` of the `StepContainer` when we need it to be sticky. We can also update all of the usages to import from `@automatitc/components` in the future.
* Make `ActionButtons` sticky on the desktop
* Make `ActionButtons` sticky with white padding-bottom only when the target positioner is out of the viewport. I also remove the `StickyFooter` to share the same sticky state.
* Make thumbnails to be sticky via CSS
* Make the browser header of the preview toolbar sticky.
* Using`showGeneratedDesigns` for `shouldStickyNavButtons` and `isSticky` for `hasStickyNavButtonsPadding` might be confusing, is there any suggestions for naming 🤔

Note that you can compare the result with https://github.com/Automattic/wp-calypso/pull/64110 which makes the scrolling behavior inside the ifame.

https://user-images.githubusercontent.com/13596067/170662916-9d9d0b3e-4abf-4a46-86c5-489bb4f9cde6.mov

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to generated design picker: `/setup/designSetup?siteSlug=<your_site>`
* Scroll down to the navbar, thumbnails, browser header, and continue button are sticky
* Resize and scroll up and down again to check the UI won't be broken
* Switch to another design and scroll up and down to check the UI won't be broken

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/63794

arthur791004 committed:
Sticky navbar and thumbnails

arthur791004 committed:
Use StickyPositioner to handle sticky padding

arthur791004 committed:
Sticky the browser header of the preview toolbar

arthur791004 committed:
Check isMobile

roo2 commented:
hmm I find that on scrolling down there is an animation as the sticky footer appears, it seems to slide in. However when I scroll back up, the sticky footer disapears immediately and there is no animation. I think we should have a similar "slide out" animation

https://user-images.githubusercontent.com/22446385/170907196-6ad42dfe-ace9-4152-80a6-01e48abfccb9.mov

roo2 commented:
Also one other minor thing with the footer, see how at the end of the page, the user is allowed to scroll a little more and the layout is a little distorted. I think at the bottom of the scroll range, the page should look the same as it does while scrolling. (i.e. the "Continue" button shouldn't move relative to the "View More Options" button)

Uploading Screen Recording 2022-05-30 at 12.42.06 pm.mov…

arthur791004 committed:
Remove margin-bottom

arthur791004 committed:
Adjust z-index of the footer

arthur791004 commented:
@roo2 Nice catch! I've fixed those 2 issues, thanks 👍

arthur791004 commented on client/landing/stepper/declarative-flow/internals/steps-repository/design-setup/style.scss:
```diff
@@ -410,11 +414,11 @@ $design-button-primary-color: rgb( 17, 122, 201 );
 			&.is-visible {
 				opacity: 1;
 				transform: translateY( 0 );
-				z-index: 1;
```
@taipeicoder Do you have any concerns to set the z-index when it's visible? If we only apply the z-index when it's visible, we won't see the slide-out animations and it will disappear immediately as it's covered by the previews.

taipeicoder replied:
Just tested and didn't break anything. I think I just added z-index: 1 only when visible because the footer would block the thumbnail section. But since we're using `pointer-events` to solve that, we can always have the footer as z-index: 1.

taipeicoder commented:
Thanks for making all these improvements @arthur791004! One thing I noticed is that there is some weird behavior with stickiness when the preview is not yet loaded/loading. Do you know what's the cause of that? See recording:

https://user-images.githubusercontent.com/797888/170918622-74dd0021-5f85-44ca-b98e-461e4139e6b3.mp4

taipeicoder commented:
> Thanks for making all these improvements @arthur791004! One thing I noticed is that there is some weird behavior with stickiness when the preview is not yet loaded/loading. Do you know what's the cause of that? See recording:
> 

The problem seems to be that `StickyPositioner` keeps changing the sticky state when components are switched to "sticky" and when the layout is recalculated the target intersects again.

arthur791004 committed:
Fix weird behavior as the height of the element changed when sticky

arthur791004 commented:
> Thanks for making all these improvements @arthur791004! One thing I noticed is that there is some weird behavior with stickiness when the preview is not yet loaded/loading. Do you know what's the cause of that? See recording:
> 
>  Screen.Recording.on.2022-05-30.at.12_40_39.mp4

The root cause is the height of the **browser header** changes according to the sticky state. However, I use className and inline style in 2 places and it leads to a little asynchronous. For example, the element is set to `fixed` but the height is set in the next frame so that the sticky state switches back and forth. I've fixed the problem by using an inline style to avoid any asynchronous.

Also, I found that the height of the footer also changed when switching to sticky as we use `translateY`. So, I made a change  to fix this behavior too to avoid the preview content bouncing down when the footer is sticky.

taipeicoder commented on client/landing/stepper/declarative-flow/internals/steps-repository/design-setup/style.scss:
```diff
@@ -385,50 +391,48 @@ $design-button-primary-color: rgb( 17, 122, 201 );
-			}
+			height: $design-setup-footer-height;
+			opacity: 0;
+			transform: translateY( 118px );
```
Should we also use `$design-setup-footer-height` here?

arthur791004 replied:
Oh yes, I forgot to change it. I think we can just use `100%`

taipeicoder commented on client/landing/stepper/declarative-flow/internals/steps-repository/design-setup/preview-toolbar.tsx:
```diff
@@ -30,6 +41,41 @@ const DesignPickerPreviewToolbar = ( {
+
+			setStickyStyle( {
+				position: 'fixed',
+				top: '109px',
```
Can we add a comment about what this magic number refers to?

arthur791004 replied:
Okay

taipeicoder commented on packages/onboarding/src/step-container/style.scss:
```diff
@@ -193,14 +193,42 @@
 
+		.step-container__navigation-logo {
+			display: none;
+			// Algin with wordpress logo in signup-header
```
Small typo here 🙈

arthur791004 replied:
Oops 🙈

taipeicoder commented:
Works so much smoother now 🙂
Can we add a special class name to `PreviewToolbar` component when o isolate it? Otherwise it would impact the existing design picker:

![Screen Shot 2022-05-30 at 4 22 23 PM](https://user-images.githubusercontent.com/797888/170950317-93cbc80f-3bda-416b-9d1c-09fc3609c3b8.png)

arthur791004 committed:
Address feedbacks

arthur791004 committed:
Fix preview toolbar in static design preview

arthur791004 commented:
Fixed, thanks 👍

taipeicoder reviewed (approved):
Great work on this one ❤️

arthur791004 commented:
Thanks for review 💯
